### ISSUE TITLE
OpenSMPTD inside Docker: "machine does not have a hostname set"

### ISSUE THREAD

**Issue by codethief** [OpenSMTPD/OpenSMTPD#1182]
I'm trying to run OpenSMTPD v6.8.0p2 inside a Docker container. Every time I fire up the container, though, it exists right away with the following (verbose) output:

```
debug: init ssl-tree
debug: init ca-tree
debug: init ssl-tree
debug: using "fs" queue backend
debug: using "ramqueue" scheduler backend
debug: using "ram" stat backend
smtpd: machine does not have a hostname set
```

Here are the relevant files:

Dockerfile:

```
FROM ubuntu:latest

RUN apt-get update
RUN apt-get install --yes \
    opensmtpd

RUN rm -rf /etc/mail

COPY smtpd.conf /etc/mail/smtpd.conf
COPY mailname /etc/mail/mailname

COPY run.sh /run/run.sh

VOLUME /volumes/aliases
VOLUME /volumes/certs
VOLUME /var/log

EXPOSE 25 465 587

CMD ["/run/run.sh"]
```

(Note: The file `mailname` contains the string "my.hostname.com" and I'm also passing `my.hostname.com` as `hostname` setting to docker-compose when running the container.)

smtpd.conf:

```
pki my.hostname.com cert "/volumes/certs/live/my.hostname.com/fullchain.pem"
pki my.hostname.com key "/volumes/certs/live/my.hostname.com/privkey.pem"

filter check_dyndns phase connect match rdns regex { '.*\.dyn\..*', '.*\.dsl\..*' } \
    disconnect "550 no residential connections"

filter check_rdns phase connect match !rdns \
    disconnect "550 no reverse DNS"

filter check_fcrdns phase connect match !fcrdns \
    disconnect "550 no forward-confirmed reverse DNS"

table virtual_users file:/volumes/aliases/virtual_users.conf
table virtual_domains file:/volumes/aliases/virtual_domains.conf
table credentials file:/volumes/aliases/credentials.conf

listen on all hostname my.hostname.com tls pki my.hostname.com
    filter { check_dyndns, check_rdns, check_fcrdns }

listen on all port submission hostname my.hostname.com tls-require pki my.hostname.com auth <credentials>

action "receive" virtual <virtual_users> lmtp "dovecot:24" rcpt-to
action "outbound" relay helo my.hostname.com

match from any for domain <virtual_domains> action "receive"
match from any auth for any action "outbound"
```

run.sh:

```bash
#!/bin/sh

set -e

# `cat /etc/mail/mailname`, `cat /etc/hostname`, `echo $HOSTNAME` and `$ hostname` all output "my.hostname.com" here.

/usr/sbin/smtpd -dv
```

I'm really at a loss as to what's happening here. As indicated above, the hostname seems to be set correctly in all possible ways. Moreover, I'm providing a `hostname` parameter in all relevant places in `smtpd.conf`, so OpenSMPTD should simply use that:

> [hostname](https://man.openbsd.org/smtpd.conf#hostname) _hostname_
>    Use hostname in the greeting banner instead of the default server name.

Finally, even if I didn't set `hostname` in `smtpd.conf`, I don't even understand why OpenSMTPD would try to determine the hostname from the system settings(?) to begin with, instead of simply looking at `/etc/mail/mailname`. After all the [man page](https://man.openbsd.org/smtpd.conf) for `smtpd.conf` says:

> /etc/mail/mailname
>    If this file exists, the first line is used as the server name. Otherwise, the server name is derived from the local hostname returned by [gethostname(3)](https://man.openbsd.org/gethostname.3), either directly if it is a fully qualified domain name, or by retrieving the associated canonical name through [getaddrinfo(3)](https://man.openbsd.org/getaddrinfo.3).

Could anyone shed light on this mystery? Thanks so much!

**Comment by dongqiuyinggrace** [comment]
I got the same issue, have you got a solution for it?

**Comment by omar-polo** [comment]
I can reproduce using 6.8.0p2 under docker but can't using the latest master.  Nothing in particular stands out in the diff between the 6.8.0p2, but it's not easy to pin-point.  Lots of time passed since 6.8.0, in the meantime there were various improvements and the rename of the processes names adds noise to the diff, I'm afraid.

Could you please try instead with the latest master and report if it worked and if any other issues were encountered?  We're getting close to being able to tag a release, so this would be really helpful.  Thanks! :)

**Comment by codethief** [comment]
@dongqiuyinggrace I have not but @omar-polo 's comment sounds promising. I'm afraid, though, I won't be able to test this in the near future due to other things on my todo list.

**Comment by omar-polo** [comment]
Hello,

We've just released OpenSMTPD-7.3.0p0, and to be sure I've rechecked this issues.  I can't reproduce it anymore.  For reference, here's the Dockerfile I'm using:


```Dockerfile
FROM alpine:latest

RUN apk add autoconf automake bison ca-certificates fts-dev gcc libevent-dev \
	libtool linux-pam-dev make musl-dev openssl-dev zlib-dev curl

RUN adduser -h /var/mail -D vmail

RUN mkdir -p /var/lib/opensmtpd/empty && \
	adduser _smtpd -h /var/lib/opensmtpd/empty/ -D -H -s /bin/false && \
	adduser _smtpq -h /var/lib/opensmtpd/empty/ -D -H -s /bin/false && \
	mkdir -p /var/spool/smtpd && \
	mkdir -p /var/mail && \
	mkdir -p /etc/mail && \
	chmod 711 /var/spool/smtpd

WORKDIR /opensmtpd
RUN curl -O https://www.opensmtpd.org/archives/opensmtpd-7.3.0p0.tar.gz && \
	tar xzf opensmtpd-7.3.0p0.tar.gz && \
	cd opensmtpd-7.3.0p0 && \
	./configure --with-gnu-ld --sysconfdir=/etc/mail --with-auth-pam \
		--with-bundled-libtls && \
	make && \
	make install && \
	cp etc/aliases /etc/mail/aliases

WORKDIR /libopensmtpd
RUN curl https://distfiles.sigtrap.nl/libopensmtpd-0.7.tar.gz | tar xzf - && \
	cd libopensmtpd-0.7 && \
	make -f Makefile.gnu all install

WORKDIR /filter-dkimsign
RUN curl https://distfiles.sigtrap.nl/filter-dkimsign-0.5.tar.gz | tar xzf - && \
	cd filter-dkimsign-0.5 && \
	make -f Makefile.gnu all install

WORKDIR /

RUN mkdir -p /etc/ssl/mail.example.com
RUN mkdir -p /etc/dkim

EXPOSE 25
EXPOSE 465
```

it's quite long but only because it also builds the dkimsign filter and adds a local vmail user needed by my specific setup.

The matching configuration:
```
pki mail.example.com cert "/etc/ssl/mail.example.com/mail.example.com.crt"
pki mail.example.com key  "/etc/ssl/mail.example.com/mail.example.com.key"

filter check_dyndns phase connect match rdns \
	regex { '.*\.dyn\..*', '.*\.dsl\..*' } \
	disconnect "550 no residential connections"

filter check_rdns phase connect match !rdns \
	disconnect "550 missing rDNS"

filter check_fcrdns phase connect match !rdns \
	disconnect "550 missing FCrDNS"

filter dkimsign \
	proc-exec "/usr/libexec/opensmtpd/filter-dkimsign -d example.com -k /etc/dkim/mail.example.com.key -s YYYYMMDD"

table aliases	file:/etc/mail/aliases
table domains	file:/etc/mail/domains
table secrets	file:/etc/mail/passwd
table virtuals	file:/etc/mail/virtuals

listen on 0.0.0.0 tls pki mail.example.com hostname mail.example.com \
	filter { check_dyndns, check_rdns, check_fcrdns, dkimsign }

listen on 0.0.0.0 port 465 tls-require pki mail.example.com \
	auth <secrets> filter dkimsign

action "remote_mail" lmtp "dovecot:24" rcpt-to virtual <virtuals>
action "local_mail" lmtp "dovecot:24" rcpt-to alias <aliases>
action "outbound" relay helo mail.example.com

match from any for domain <domains> action "remote_mail"
match for local action "local_mail"

match from any auth for any action "outbound"
match for any action "outbound"
```

if I omit `hostname` it just picks up the id of the container somehow.  I'm running this on a Rocky Linux installation using podman (running docker on redhat-adjacent distro is quite hard!)

HTH

**Comment by omar-polo** [comment]
I've been running for months OpenSMTPD under podman, so I'm going to close this issue.

By the way, now we provide [docker images for releases](https://github.com/OpenSMTPD/OpenSMTPD/pkgs/container/opensmtpd) via the github registry too.
